doc/03_research/03_planning/04_decision_making.md

```diff
@@ -0,0 +1,256 @@
+
+#### Decision tree
+
+* This team used a decision tree to cover the major driving scenarios
+* Within the scenarios the actions are more linear
+* Reminds me of the execution of a state where driving scenarios are the states and the execution the things our local planner should do within that state
+
+#### Take Away
+
+* Even though the approach is different, the execution might be similar to the other team algorithms
+* We might not be interested in a decision tree as we want to keep the option to switch to a Markov chain, which would be more overhead if we start with a decision tree
+
+## Python or ROS libraries for these decision-making algorithms
+
+### State machines
+
+#### SMACH
+
+* Task-level architecture for creating state machines for robot behaviour.
+* Based on Python
+* Fast prototyping: Quickly create state machines
+* Complex state machines can easily be created
+* Introspection: smach_viewer provides a visual aid to follow the state machine executing its tasks
+  * smach_viewer is unmaintained and does not work with noetic
+* Allows nested state machines
+* Values can be passed between states
+* Tutorials and documentation seems to be easy to understand so creating a first state machine shouldn't be too hard
+* working with several ROS topics and messages within the state machine needs to be evaluated:
+  * the execution of states is mostly planned to happen in the local planner so for just sending a ROS message, SMACH might be efficient
+
+Not use SMACH for:
+
+* Unstructured tasks: SMACH is not efficient in sheduling unstructured tasks
+* Low-level systems: SMACH is not build for high efficiency, might fall short for emergency maneuvers
+
+* Simple examples run without problem
+
+#### SMACC
+
+* event-driven, asynchronous, behavioral state machine library
+* real-time ROS applications
+* written in C++
+* designed to allow programmers to build robot control applications for multicomponent robots, in an intuitive and systematic manner.
+* well maintained, lots of prebuild state machines to possibly start from
+
+Why not use SMACC:
+
+* might get some time to get back into C++
+* more sophisticated library might need more time to get used to
+* awful country music in the back of tutorial videos
+
+* Tutorials do not run without further debugging which I didn't invest the time to do so
+
+### Markov Chains
+
+#### QuantEcon
+
+* a economics library for implementing Markov chains
+* more focussed on simulation than actually using it in an AD agent
+* maybe usable for testing and simulating a Markov chain before implementing it
+
+#### markov_decision_making
+
+* ROS library for robot decision-making based on Markov Decision Problems
+* written in C++
+* callback-based action interpretation allows to use other frameworks (SMACH)
+* relatively easy to implement hierarchical MDPs
+* supports synchronous and asynchronous execution
+
+Why not use markov_decision_making:
+
+* not maintained
+* only works with ROS hydro
```

Comment:
Wie problematisch ist das? Können wir das zum laufen bringen oder kommt markov_decision_making daher eigl nicht in Frage?

Reply:
Ich konnte die Beispiele nicht zum Laufen bringen und bin auch bisher auf keine weitere Implementierung gestoßen, ich habe aber einige selbstimplementierte Lösungen in Python gefunden, die gut verständlich waren.